(fix) O3-3287: Collapsible Form Pages

TrevorAntony opened the pull request:
## Requirements
Form pages should be collapsible, just as form sections are.

## Summary
This PR ensure Form pages are collapsible just like the sections

## Screenshots
<img width="1433" alt="Screenshot 2024-09-16 at 16 29 03" src="https://github.com/user-attachments/assets/71ab19ea-a5b6-4596-a69a-f0e770301aea">
<img width="1436" alt="Screenshot 2024-09-16 at 16 29 14" src="https://github.com/user-attachments/assets/73efe369-6240-4cbe-a54b-20ff1923c26e">


## Related Issue
https://openmrs.atlassian.net/browse/O3-3287

samuelmale commented on src/components/renderer/page/page.renderer.component.tsx:
```diff
@@ -7,12 +7,15 @@ import { Waypoint } from 'react-waypoint';
 
 interface PageRendererProps {
   page: FormPage;
+  isCollapsed: boolean;
+  onToggleCollapse: () => void;
```
Any reason as to why the page renderer isn't locally managing this state?

samuelmale commented on src/components/renderer/form/form-renderer.component.tsx:
```diff
@@ -33,6 +33,8 @@ export const FormRenderer = ({ processorContext, initialValues, setIsLoadingForm
     formJson: evaluatedFormJson,
   });
 
+  const [collapsedPages, setCollapsedPages] = useState<Set<string>>(new Set());
```
Any reason as to why the parent component has to manage this state?

samuelmale commented on src/hooks/useFormCollapse.tsx:
```diff
@@ -29,8 +30,22 @@ export function useFormCollapse(sessionMode: SessionMode) {
   return {
     isFormExpanded,
     hideFormCollapseToggle,
+    collapsedPages,
+    togglePageCollapse,
```
Is this being anywhere?

pirupius commented:
@TrevorAntony as u make changes to this, please adjust the UI to match the pattern for the sections with the icon being on the far right

TrevorAntony commented:
> @TrevorAntony as u make changes to this, please adjust the UI to match the pattern for the sections with the icon being on the far right

<img width="1430" alt="Screenshot 2024-09-19 at 10 38 55" src="https://github.com/user-attachments/assets/57181ef1-bde5-461c-83d9-6ad7c08eb54e">
Moved it to the right and added the hover

pirupius commented:
@TrevorAntony thanks for the good work so far. There seems to be an unnecessary padding/margin that makes the text and arrow not align. Please resolve it.

TrevorAntony commented:
> @TrevorAntony thanks for the good work so far. There seems to be an unnecessary padding/margin that makes the text and arrow not align. Please resolve it.

Thank you. Yes i have seen. Made the styling consistent now
<img width="1434" alt="Screenshot 2024-09-19 at 11 05 33" src="https://github.com/user-attachments/assets/459d8358-2029-44a4-8bd5-8de39c10def8">

pirupius commented on src/components/renderer/form/form-renderer.component.tsx:
```diff
@@ -91,7 +91,16 @@ export const FormRenderer = ({ processorContext, initialValues, setIsLoadingForm
           );
         }
-        return <PageRenderer page={page} />;
+        return (
+          <PageRenderer
+            key={page.label}
+            page={page}
+            isCollapsed={false}
+            onToggleCollapse={function (): void {
+              throw new Error('Function not implemented.');
+            }}
+          />
+        );
```
From the look of things you had added this implementation because the state was being managed within this file and now that it has been moved to the `PageRenderer`, there's no need for the extra params

pirupius commented on src/components/renderer/page/page.renderer.component.tsx:
```diff
@@ -7,40 +7,58 @@ import { Waypoint } from 'react-waypoint';
 
 interface PageRendererProps {
   page: FormPage;
+  isCollapsed: boolean;
+  onToggleCollapse: () => void;
```
Apply the same clean up here

pirupius commented on src/components/sidebar/sidebar.component.tsx:
```diff
@@ -20,6 +21,8 @@ interface SidebarProps {
   selectedPage: string;
   setValues: (values: unknown) => void;
   values: object;
+  collapsedPages: Set<string>;
+  togglePageCollapse: (pageId: string) => void;
```
Did u add some UI elements to update the page state?

pirupius replied:
For context, that component renders this section of the form engine. It was previously removed during the RHF migration. Notice how it shows only the section title and it's sole purpose to quickly scroll to a specific section of the form. That's what i mean't when i asked this. 

<img width="1029" alt="Screenshot 2024-09-19 at 14 15 31" src="https://github.com/user-attachments/assets/eb3d6e57-568a-40ef-9349-ddde33313ddd">

TrevorAntony replied:
Oh okay, i was under the wrong impression. Since that is the case there is no need to make changes to the sidebar component or styling

pirupius commented on src/components/sidebar/sidebar.component.tsx:
```diff
@@ -88,7 +95,14 @@ const Sidebar: React.FC<SidebarProps> = ({
-            <div className={styles.sectionLink}>{page.label}</div>
+            <div className={styles.sectionLink}>
+              {page.label}
+              {isCollapsed ? (
+                <ChevronUpIcon className={styles.collapseIcon} aria-label="Expand" />
+              ) : (
+                <ChevronDownIcon className={styles.collapseIcon} aria-label="Collapse" />
+              )}
```
Am having a little bit of an issue with this, The sidebar only displays the page title for quick navigation. Do u have a screenshot of what your trying to achieve here

TrevorAntony replied:
Yes, there was no UI element to show the pages are collapsible like the sections so i added the ChevronUp and Down.
Screenshot
<img width="1436" alt="Screenshot 2024-09-19 at 13 47 01" src="https://github.com/user-attachments/assets/5274f3e0-464f-47d0-be28-41a50e729ace">

pirupius replied:
See previous comment as it addresses this as well

pirupius replied:
This is a previous design with the sidenav visible https://app.zeplin.io/project/60d5947dd636aebbd63dce4c/screen/61bb008f0e69e45e7175e63e

pirupius commented on src/components/renderer/form/form-renderer.component.tsx:
```diff
@@ -1,4 +1,4 @@
-import React, { useEffect, useMemo, useReducer } from 'react';
+import React, { useEffect, useMemo, useReducer, useState } from 'react';
```
More clean up, we don't need the extra `useState` anymore
